conjure-lib/src/main/java/com/palantir/conjure/java/lib/internal/ConjureCollections.java

```diff
@@ -37,6 +39,35 @@ private ConjureCollections() {
         // cannot instantiate
     }
 
+    /*
+     * This is bizarre. Allow me to explain...
+     *
+     * We do _not_ want to expose the Conjure*List types externally
+     * but we also want the optimizations they provide to make it thru
+     * to jackson for serialization. So the runtime type needs to be
+     * preserved while also not exposing the type :phew:.
+     *
+     * To achieve this we have to do some gymnastics surrounding the type
+     * system. We need this to return the type of the list given, but also
+     * return specific Conjure types when detected. This requires that we
+     * erase the type info, but we know this is safe because we are directly
+     * returning the same type which is by definition the identity function.
+     * Therefore the input List<T> is the same types as the output List<T>.
+     */
+    public static <T> List<T> unmodifiableList(List<T> list) {
+        // Return the unmodifiable version of the Eclipse types
+        if (list instanceof ConjureIntegerList) {
+            return (List<T>) ((ConjureIntegerList) list).asUnmodifiable();
+        } else if (list instanceof ConjureDoubleList) {
+            return (List<T>) ((ConjureDoubleList) list).asUnmodifiable();
+        } else if (list instanceof ConjureSafeLongList) {
+            return (List<T>) ((ConjureSafeLongList) list).asUnmodifiable();
+        } else {
+            // Otherwise use the JDK types
+            return Collections.unmodifiableList(list);
+        }
```

Comment:
we don't need to bring along any of the implementations, they're all internal details -- `return Collections.unmodifiableList(list);` is sufficient

Reply:
Is this the only case you care about or is this just a comment on the first instance? For example, I added a number of methods to the `ConjureDoubleList` shaped types for the same reasons that we want the APIs to exist (also they are even more internal), but there isn't really a "pass-through" implementation equivalent.

Reply:
`ConjureDoubleList` is only created and checked from within `conjure-lib` code. Because it's kept insulated, we can make any changes we need without letting them soak first. In the case where code runs with a faulty classloader and inadvertently forces the conjure-lib version downward, they may not receive optimizations yet, but nothing would fail because the public API surface still exists.

Reply:
I understand that this isn't a concern within a jar, but adding API's like https://github.com/palantir/conjure-java/pull/2417/files/bc1e6311d872a0e808367c373369fd0a3d9aff2f#diff-c7c23a70bf9b00504cc05e26d2bb08a161f4e036eadd36cd0e13efbbddb4e02cR198 is necessary for the reasons you laid out. Now I could have that just call https://github.com/palantir/conjure-java/pull/2417/files/bc1e6311d872a0e808367c373369fd0a3d9aff2f#diff-c7c23a70bf9b00504cc05e26d2bb08a161f4e036eadd36cd0e13efbbddb4e02cR202 but I am trying to tease out your goal. Is it to minimize the surface area of this change?

Reply:
Understood, will slim as much as possible